File: csharp/Pattern.cs
// Access pattern detector (Adaptive-prefetch-spec §4).

using System.Collections.Generic;

namespace Nxs;

public enum AccessPattern
{
    Unknown,
    Sequential,
    Random,
    Mixed,
}

public static class PatternConstants
{
    public const int SequentialThreshold = 10;
    public const int RandomThreshold = 100;
    public const int HistorySize = 32;
    public const int MinObservations = 8;
    public const int UpgradeSequentialThreshold = 100;
}

/// <summary>Observes record indices and classifies access patterns.</summary>
public sealed class AccessPatternDetector
{
    private readonly long[] _accesses = new long[PatternConstants.HistorySize];
    private int _writePos;
    private int _filled;
    private uint _sequentialRuns;
    private uint _randomJumps;
    private long _lastIndex = -1;

    public AccessPatternDetector()
    {
        for (int i = 0; i < _accesses.Length; i++)
            _accesses[i] = -1;
    }

    public uint SequentialRuns => _sequentialRuns;

    public long LastIndex => _lastIndex;

    public void Observe(int index)
    {
        long idx = index;
        if (_lastIndex >= 0)
        {
            ulong delta = idx >= _lastIndex
                ? (ulong)(idx - _lastIndex)
                : (ulong)(_lastIndex - idx);
            if (delta <= PatternConstants.SequentialThreshold)
            {
                if (_sequentialRuns < uint.MaxValue)
                    _sequentialRuns++;
            }
            else if (delta > PatternConstants.RandomThreshold)
            {
                if (_randomJumps < uint.MaxValue)
                    _randomJumps++;
            }
        }
        _accesses[_writePos] = idx;
        _writePos = (_writePos + 1) % PatternConstants.HistorySize;
        if (_filled < PatternConstants.HistorySize)
            _filled++;
        _lastIndex = idx;
    }

    public AccessPattern Pattern()
    {
        int total = (int)_sequentialRuns + (int)_randomJumps;
        if (total < PatternConstants.MinObservations)
            return AccessPattern.Unknown;
        if (_sequentialRuns > _randomJumps * 3)
            return AccessPattern.Sequential;
        if (_randomJumps > _sequentialRuns)
            return AccessPattern.Random;
        return AccessPattern.Mixed;
    }

    public static string PatternName(AccessPattern p) => p switch
    {
        AccessPattern.Sequential => "sequential",
        AccessPattern.Random => "random",
        AccessPattern.Mixed => "mixed",
        _ => "unknown",
    };

    /// <summary>Predicted next record indices when pattern is sequential (§4.4).</summary>
    public List<int> PredictNext(int depth, int recordCount)
    {
        if (Pattern() != AccessPattern.Sequential || _lastIndex < 0)
            return new List<int>();
        int start = (int)_lastIndex + 1;
        var outList = new List<int>(depth);
        for (int i = 0; i < depth; i++)
        {
            int idx = start + i;
            if (idx < recordCount)
                outList.Add(idx);
        }
        return outList;
    }
}

File: php/Pattern.php
<?php
/**
 * Access pattern detector (Adaptive-prefetch-spec §4).
 */

namespace Nxs;

const SEQUENTIAL_THRESHOLD = 10;
const RANDOM_THRESHOLD = 100;
const HISTORY_SIZE = 32;
const MIN_OBSERVATIONS = 8;
const UPGRADE_SEQUENTIAL_THRESHOLD = 100;

const PATTERN_UNKNOWN = 'unknown';
const PATTERN_SEQUENTIAL = 'sequential';
const PATTERN_RANDOM = 'random';
const PATTERN_MIXED = 'mixed';

const DEFAULT_PREFETCH_DEPTH = 4;
const EAGER_THRESHOLD_MB = 10;
const LAZY_THRESHOLD_MB = 50;

/**
 * Observes record indices and classifies access patterns.
 */
class AccessPatternDetector
{
    /** @var int[] */
    private array $accesses;
    private int $writePos = 0;
    private int $filled = 0;
    public int $sequentialRuns = 0;
    private int $randomJumps = 0;
    public int $lastIndex = -1;

    public function __construct()
    {
        $this->accesses = array_fill(0, HISTORY_SIZE, -1);
    }

    public function observe(int $index): void
    {
        if ($this->lastIndex >= 0) {
            $delta = abs($index - $this->lastIndex);
            if ($delta <= SEQUENTIAL_THRESHOLD) {
                $this->sequentialRuns = min($this->sequentialRuns + 1, PHP_INT_MAX);
            } elseif ($delta > RANDOM_THRESHOLD) {
                $this->randomJumps = min($this->randomJumps + 1, PHP_INT_MAX);
            }
        }
        $this->accesses[$this->writePos] = $index;
        $this->writePos = ($this->writePos + 1) % HISTORY_SIZE;
        if ($this->filled < HISTORY_SIZE) {
            $this->filled++;
        }
        $this->lastIndex = $index;
    }

    public function pattern(): string
    {
        $total = $this->sequentialRuns + $this->randomJumps;
        if ($total < MIN_OBSERVATIONS) {
            return PATTERN_UNKNOWN;
        }
        if ($this->sequentialRuns > $this->randomJumps * 3) {
            return PATTERN_SEQUENTIAL;
        }
        if ($this->randomJumps > $this->sequentialRuns) {
            return PATTERN_RANDOM;
        }
        return PATTERN_MIXED;
    }

    /** @return int[] */
    public function predictNext(int $depth, int $recordCount): array
    {
        if ($this->pattern() !== PATTERN_SEQUENTIAL || $this->lastIndex < 0) {
            return [];
        }
        $start = $this->lastIndex + 1;
        $out = [];
        for ($i = 0; $i < $depth; $i++) {
            $idx = $start + $i;
            if ($idx < $recordCount) {
                $out[] = $idx;
            }
        }
        return $out;
    }
}

/**
 * Initial prefetch strategy from hint and file size (§5.1).
 */
function initialPrefetchStrategy(int $hint, int $fileSize): string
{
    $fileSizeMb = intdiv($fileSize, 1024 * 1024);
    if ($hint === HINT_FULL && $fileSizeMb <= EAGER_THRESHOLD_MB) {
        return 'eager';
    }
    if ($fileSizeMb > LAZY_THRESHOLD_MB) {
        return 'lazy';
    }
    return 'adaptive';
}

/** @return array{0: int, 1: int} row data sector [start, length] */
function rowDataSector(int $tailStart, int $fileSize): array
{
    $sectorStart = 32;
    if ($tailStart > $sectorStart && $tailStart <= $fileSize) {
        return [$sectorStart, $tailStart - $sectorStart];
    }
    return [$sectorStart, 0];
}
